pkg/kubelet/kubelet_node_status.go

```diff
@@ -735,22 +735,40 @@ func (kl *Kubelet) setNodeReadyCondition(node *v1.Node) {
 	// This is due to an issue with version skewed kubelet and master components.
 	// ref: https://github.com/kubernetes/kubernetes/issues/16961
 	currentTime := metav1.NewTime(kl.clock.Now())
-	var newNodeReadyCondition v1.NodeCondition
+	newNodeReadyCondition := v1.NodeCondition{
+		Type:              v1.NodeReady,
+		Status:            v1.ConditionTrue,
+		Reason:            "KubeletReady",
+		Message:           "kubelet is posting ready status",
+		LastHeartbeatTime: currentTime,
+	}
 	rs := append(kl.runtimeState.runtimeErrors(), kl.runtimeState.networkErrors()...)
-	if len(rs) == 0 {
+	if len(rs) > 0 {
 		newNodeReadyCondition = v1.NodeCondition{
 			Type:              v1.NodeReady,
-			Status:            v1.ConditionTrue,
-			Reason:            "KubeletReady",
-			Message:           "kubelet is posting ready status",
+			Status:            v1.ConditionFalse,
+			Reason:            "KubeletNotReady",
+			Message:           strings.Join(rs, ","),
 			LastHeartbeatTime: currentTime,
 		}
-	} else {
+	}
+	requiredCapacities := []v1.ResourceName{v1.ResourceCPU, v1.ResourceMemory, v1.ResourcePods}
+	if utilfeature.DefaultFeatureGate.Enabled(features.LocalStorageCapacityIsolation) {
+		requiredCapacities = append(requiredCapacities, v1.ResourceEphemeralStorage)
+	}
+	missingCapacities := []string{}
+	for _, resource := range requiredCapacities {
+		if _, found := node.Status.Capacity[resource]; !found {
+			missingCapacities = append(missingCapacities, string(resource))
+		}
+	}
+	if len(missingCapacities) > 0 {
+		glog.Warningf("Node is NotReady because we are missing capacity information for resources: %v", missingCapacities)
 		newNodeReadyCondition = v1.NodeCondition{
 			Type:              v1.NodeReady,
 			Status:            v1.ConditionFalse,
 			Reason:            "KubeletNotReady",
-			Message:           strings.Join(rs, ","),
+			Message:           fmt.Sprintf("Missing node capacity for resources: %s", strings.Join(missingCapacities, ", ")),
```

Comment:
I think we should append this message to the previous message.

Reply:
Eh, but the previous one could be "kubelet is posting ready status".  I think surfacing one is fine.  It is probably more useful to have the runtime errors override the resource capacity errors anyways.

Reply:
spoke offline.  Just append to rs list